l3gd20h migration to HAL

WilliamLim781 opened the pull request:
Migrated igyroscope.hpp, l3gd20h.hpp, and l3gd20h.cpp to HAL. Most of the changes had to do with WriteReg, ReadReg, and Receive since these functions are related to I2C communication.

### Summary

- created a new folder under the sensor folder called GyroScope and made igyroscope.hpp, l3gd20h.hpp, and l3gd20h.cpp files in that folder
- altered the functions WriteReg, ReadReg, Receive to operate with native HAL I2C functions on stm32
- removed the adapter namespace (noticed you didn't have it on linear potentiometers so i removed them here too)
- copied the comment format that the HAL uses to describe a function use and the parameters used in a function 
![i2c example](https://github.com/DallasFormulaRacing/DataAcquisition2.0/assets/113472670/c22e9589-581c-4e0f-9449-572dfc3ae92a)
this is an example of the format that I took from the stm32 HAL I2C library.  I was just curious how the abstract interface would look like having some sort of standard commenting. let me know what you think about this comment format.

### Remarks

When migrating the abstract interface to HAL I noticed that both the GetDegreesPerSecond() and GetRadiansPerSecond() functions used the angular velocity array without calling the ComputeAngularVelocity() function. So when I tried using the functions I was just getting zeros in the serial monitor. So now I added the ComputeAngularVelocity() function to both to make sure that angular velocity is computed/obtained before these functions do any conversion to either degrees/second or radians/second.  Not really sure how the functions worked before so maybe I'm missing something but regardless I added the ComputeAngularVelocity() function and everything is functioning as expected so that's cool. The picture is just showing where I added the ComputeAngularVelocity() function.

![i2c function](https://github.com/DallasFormulaRacing/DataAcquisition2.0/assets/113472670/6b04c5c1-080f-4068-94f1-3755fa4f5558)

Finally, I wasn't sure how to use the unique_ptr that was used on the linear potentiometers so I just went with the original way of using 
IGyroscope *l3gd20h = new Gyroscope_L3GD20H();
If that needs to be changed to a unique_ptr then that can be done. Just didn't know how to do it.



### Checks
- [x] GetDegreesPerSecond functions as expected
- [x] GetRadiansPerSecond functions as expected 
- [x] ComputeAngularVelocity functions as expected
- [x] l3gd20h specific functions operate as expected

WilliamLim781 committed:
l3gd20h migration to HAL

Migrated igyroscope.hpp, l3gd20h.hpp, and l3gd20h.cpp to HAL. Most of the changes had to do with WriteReg, ReadReg, and Receive since these functions are related to I2C communication.

crsz20 commented on Project/DAQ_System/Core/Src/i2c.c:
```diff
@@ -0,0 +1,130 @@
+  /* USER CODE BEGIN I2C1_Init 1 */
+
+  /* USER CODE END I2C1_Init 1 */
+  hi2c1.Instance = I2C1;
+  hi2c1.Init.ClockSpeed = 100000;
+  hi2c1.Init.DutyCycle = I2C_DUTYCYCLE_2;
+  hi2c1.Init.OwnAddress1 = 0;
+  hi2c1.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
+  hi2c1.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+  hi2c1.Init.OwnAddress2 = 0;
+  hi2c1.Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
+  hi2c1.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
```
Consider documenting your CubeMX `.ioc` configs on the [wiki page](https://github.com/DallasFormulaRacing/DataAcquisition2.0/wiki/Accelerometer).

crsz20 commented on Project/DAQ_System/Program/Src/app.cpp:
```diff
@@ -22,25 +22,33 @@ extern UART_HandleTypeDef huart3;
 int counter = 0;
 uint32_t TxMailbox;
 
+// Gyroscope variables
+short *GyroData; // first element is X-axis, the second element is the Y-axis, and the third element is the z-axis.
```
Move this to the inside of `cppMain()`

crsz20 commented on Project/DAQ_System/Program/Src/app.cpp:
```diff
@@ -93,6 +104,9 @@ void cppMain() {
 	std::unique_ptr<sensor::ILinearPotentiometer> lin_pot(nullptr);
 	lin_pot = std::make_unique<sensor::SLS1322>(hadc1);
 
+	IGyroscope *l3gd20h = new Gyroscope_L3GD20H();
+
```
Great! This is exactly what we need. However, raw pointers require manual deletion. Also, for our use case, we do not want multiple pointers (aka multiple, or shared references) to the object being pointed to.

Smart Pointers are a thin "wrapper" class. In other words, smart pointers are C++ objects that have a **raw** pointer inside it — so they are a little bit nicer to work with and do not require manual deletion of the object.

`unique_ptr` is a smart pointer that allows only one pointer to the object.

To use it, try the following:
```c++
std::unique_ptr<sensor::IGyroscope> gyroscope(nullptr);      // Declare an abstract pointer
gyroscope = std::make_unique<sensor::L3GD20H>(hi2c1);   // Create an object and assign it to the pointer
```

Notice that I named the pointer `gyroscope` instead of `l3gd20h`. This is because the abstract interface pointer should not care about which gyroscope is being used, as long as it satisfies the rules of the `IGyroscope` interface.

crsz20 commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.hpp:
```diff
@@ -0,0 +1,62 @@
+
+
+
+class Gyroscope_L3GD20H: public IGyroscope{
```
Rename the class from `Gyroscope_L3GD20H` to simply `L3GD20H`. The team had mutually decided to shorten the name as it is already reflected that this is a gyroscope through the folder directory, the interface class, and later in our docs.

crsz20 commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/igyroscope.hpp:
```diff
@@ -0,0 +1,32 @@
+
+        virtual short* GetDegreesPerSecond() = 0;
+        virtual float* GetRadiansPerSecond() = 0;
+        virtual bool ComputeAngularVelocity() = 0;
```
If `ComputeAngularVelocity()` is no longer required to be used by the user, make it `private` instead of `public`

crsz20 commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.cpp:
```diff
@@ -0,0 +1,150 @@
+#include "l3gd20h.hpp"
+#include <stdint.h>
+
+#include "i2c.h"
+extern I2C_HandleTypeDef hi2c1;
+
+
+
+Gyroscope_L3GD20H::Gyroscope_L3GD20H(){
```
Consider requiring the `hi2c1` object to be a parameter of the class constructor. This is called "dependency injection".

You can look at the linpots for a quick example:
* [Included the auto-generated](https://github.com/DallasFormulaRacing/DataAcquisition2.0/blob/develop/Project/DAQ_System/Program/Src/Sensor/LinearPotentiometer/sls1322.hpp#L17) ADC header file
* [Declared an ADC object](https://github.com/DallasFormulaRacing/DataAcquisition2.0/blob/develop/Project/DAQ_System/Program/Src/Sensor/LinearPotentiometer/sls1322.hpp#L58) in the header file.
* [Required the object to be passed through the constructor](https://github.com/DallasFormulaRacing/DataAcquisition2.0/blob/develop/Project/DAQ_System/Program/Src/Sensor/LinearPotentiometer/sls1322.cpp#L18) and initialized it to the object we declared. The part after the `:` is the "initialization list".
* Then, in practice, the linpot is [initialized like this](https://github.com/DallasFormulaRacing/DataAcquisition2.0/blob/develop/Project/DAQ_System/Program/Src/app.cpp#L94).

WilliamLim781 committed:
resolving github conversation recommendation

- moved *GyroData to cppmain()
- changed to smart pointer
- renamed gyro class from Gyroscope_L3GD20H to L3GD20H
- made ComputeAngularVelocity() into a private function
- added he dependency injection to the constructor

crsz20 commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.hpp:
```diff
@@ -0,0 +1,105 @@
+        short offset_average_[3] = {0};
+
+        /// The I2C peripheral from ST's HAL library.
+        I2C_HandleTypeDef& hi2c_;
```
I like to remove the "h" to make it more distinct from HAL's `hi2c` object. To me, the "h" kind of stands for HAL but it's also a practice in C to call an object/pointer/interface a "handle".

Let me know what you think! Perhaps something to add to our style guide.

WilliamLim781 replied:
I like the idea of removing the "h" but at the same time, I think the "h" might be useful for newer members and those who aren't familiar with the HAL. As you said, it's a good practice to have the "h" to dictate we are using a handle. So keeping it can serve as a good reference for other members to learn good practices as they move on to make their own libraries and such.  in the end, it's honestly up to you, I don't mind either way.

crsz20 reviewed (approved):
Awesome work!

WilliamLim781 committed:
Merge branch 'develop' into featue/138_Migrate-L3GD20H-to-HAL

crsz20 commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.cpp:
```diff
@@ -0,0 +1,122 @@
+		return false;
+	}
+	return true;
+}
```
`ReadReg()` seems to be unused.

crsz20 commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/igyroscope.hpp:
```diff
@@ -0,0 +1,34 @@
+        virtual short* GetDegreesPerSecond() = 0;
+        virtual float* GetRadiansPerSecond() = 0;
+    private:
+        virtual bool ComputeAngularVelocity() = 0;
```
Optional stylistic choice: Remove `Get` and `Compute` from the method names. The return types + our API docs should make this clear enough.

crsz20 commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.cpp:
```diff
@@ -0,0 +1,122 @@
+#include "l3gd20h.hpp"
+#include <stdint.h>
+
+#include "i2c.h"
```
Remove duplicate `include`

WilliamLim781 committed:
gyro code fixes

- removed the "get" and "compute " from the function names.
- removed ReadReg() function since it wasn't used
- removed the duplicate "I2c.h" library

WilliamLim781 committed:
bug fixes

had to remove all instances of "get" and "compute" from the other files that weren't the abstract interface.

dani4ko03 commented:
What was the comment format before you copied it from HAL?

WilliamLim781 commented:
> What was the comment format before you copied it from HAL?

I don't recall there being a general format for comments before the migration to HAL. There were just comments in places that seemed confusing or could be potentially confusing to people who were reading the library.

dani4ko03 commented:
> > What was the comment format before you copied it from HAL?
> 
> I don't recall there being a general format for comments before the migration to HAL. There were just comments in places that seemed confusing or could be potentially confusing to people who were reading the library.

I was just curious. This way of commenting looks good to me thought.

dani4ko03 reviewed (approved):
Looks great to me!

WilliamLim781 committed:
resolved

alexshidagoatnocap commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.cpp:
```diff
@@ -0,0 +1,111 @@
+#include <stdint.h>
+
+
+namespace sensor{
```
Add a space. Change to `sensor {`

alexshidagoatnocap commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.cpp:
```diff
@@ -0,0 +1,111 @@
+
+namespace sensor{
+
+L3GD20H::L3GD20H(I2C_HandleTypeDef& hi2c):hi2c_(hi2c){
```
Add a space. Change to `L3GD20H::L3GD20H(I2C_HandleTypeDef& hi2c):hi2c_(hi2c) {`

alexshidagoatnocap commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.cpp:
```diff
@@ -0,0 +1,111 @@
+    return radians_per_second_;
+}
+
+void L3GD20H::ComputeInitialOffset(){
```
Add space

alexshidagoatnocap commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.cpp:
```diff
@@ -0,0 +1,111 @@
+}
+
+
+bool L3GD20H::WriteReg(uint8_t addr_i2c, uint8_t addr_reg, uint8_t register_data){
```
Add space

alexshidagoatnocap commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.cpp:
```diff
@@ -0,0 +1,111 @@
+
+
+
+bool L3GD20H::Receive(uint8_t data_buf[]){
```
Add space

alexshidagoatnocap commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.cpp:
```diff
@@ -0,0 +1,111 @@
+
+bool L3GD20H::Receive(uint8_t data_buf[]){
+	uint8_t addr_reg = L3GD20_OUT_X_L | 0x80; // to set sequential reading you need to make the MSB of the register address to 1;
+	if(HAL_I2C_Master_Transmit(&hi2c1, GYR_ADDRESS, &addr_reg, 1, HAL_MAX_DELAY) != HAL_OK){
```
Add space. Change to `if (HAL_I2C_Master_Transmit`

alexshidagoatnocap commented on Project/DAQ_System/Program/Src/Sensor/GyroScope/l3gd20h.cpp:
```diff
@@ -0,0 +1,111 @@
+	if(HAL_I2C_Master_Transmit(&hi2c1, GYR_ADDRESS, &addr_reg, 1, HAL_MAX_DELAY) != HAL_OK){
+		return false;
+	}
+	if(HAL_I2C_Master_Receive(&hi2c1, GYR_ADDRESS, data_buf, 6, HAL_MAX_DELAY) != HAL_OK){
```
Add space.

alexshidagoatnocap reviewed (commented):
Code is well written, just a few spaces are needed for consistency reasons.
